pkg/cloud/azure/services/virtualmachines/virtualmachines.go

```diff
@@ -436,6 +423,127 @@ func generateImagePlan(image machinev1.Image) *compute.Plan {
 	}
 }
 
+func generateOSDisk(vmSpec *Spec) *compute.OSDisk {
+	osDisk := &compute.OSDisk{
+		Name:         to.StringPtr(fmt.Sprintf("%s_OSDisk", vmSpec.Name)),
+		OsType:       compute.OperatingSystemTypes(vmSpec.OSDisk.OSType),
+		CreateOption: compute.DiskCreateOptionTypesFromImage,
+		ManagedDisk:  &compute.ManagedDiskParameters{},
+		DiskSizeGB:   to.Int32Ptr(vmSpec.OSDisk.DiskSizeGB),
+	}
+
+	if vmSpec.OSDisk.ManagedDisk.StorageAccountType != "" {
+		osDisk.ManagedDisk.StorageAccountType = compute.StorageAccountTypes(vmSpec.OSDisk.ManagedDisk.StorageAccountType)
+	}
+	if vmSpec.OSDisk.ManagedDisk.DiskEncryptionSet != nil {
+		osDisk.ManagedDisk.DiskEncryptionSet = &compute.DiskEncryptionSetParameters{ID: to.StringPtr(vmSpec.OSDisk.ManagedDisk.DiskEncryptionSet.ID)}
+	}
+	if vmSpec.OSDisk.ManagedDisk.SecurityProfile.SecurityEncryptionType != "" {
+		osDisk.ManagedDisk.SecurityProfile = &compute.VMDiskSecurityProfile{}
+
+		osDisk.ManagedDisk.SecurityProfile.SecurityEncryptionType = compute.SecurityEncryptionTypes(string(vmSpec.OSDisk.ManagedDisk.SecurityProfile.SecurityEncryptionType))
+
+		if vmSpec.OSDisk.ManagedDisk.SecurityProfile.DiskEncryptionSet.ID != "" {
+			osDisk.ManagedDisk.SecurityProfile.DiskEncryptionSet = &compute.DiskEncryptionSetParameters{ID: pointer.String(vmSpec.OSDisk.ManagedDisk.SecurityProfile.DiskEncryptionSet.ID)}
+		}
+	}
+
+	return osDisk
+}
+
+func generateSecurityProfile(vmSpec *Spec, osDisk *compute.OSDisk) (*compute.SecurityProfile, error) {
+	if vmSpec.SecurityProfile == nil {
+		return nil, nil
+	}
+
+	securityProfile := &compute.SecurityProfile{
+		EncryptionAtHost: vmSpec.SecurityProfile.EncryptionAtHost,
+	}
+
+	if osDisk.ManagedDisk != nil &&
+		osDisk.ManagedDisk.SecurityProfile != nil &&
+		osDisk.ManagedDisk.SecurityProfile.SecurityEncryptionType != "" {
+
+		if vmSpec.SecurityProfile.SecurityType != machinev1.SecurityTypesConfidentialVM {
+			return nil, apierrors.InvalidMachineConfiguration("failed to generate security profile for vm %s. "+
+				"SecurityType should be set to %s when SecurityEncryptionType is defined.",
+				vmSpec.Name, compute.SecurityTypesConfidentialVM)
+		}
+
+		if vmSpec.SecurityProfile.UEFISettings == nil {
+			return nil, apierrors.InvalidMachineConfiguration("failed to generate security profile for vm %s. "+
+				"UEFISettings should be set when SecurityEncryptionType is defined.", vmSpec.Name)
+		}
+
+		if vmSpec.SecurityProfile.UEFISettings.VirtualizedTrustedPlatformModule == "" ||
+			vmSpec.SecurityProfile.UEFISettings.VirtualizedTrustedPlatformModule != machinev1.VirtualizedTrustedPlatformModulePolicyEnabled {
```

Comment:
do we need the first clause here?

could this be re-written as:
```
if vmSpec.SecurityProfile.UEFISettings.VirtualizedTrustedPlatformModule != machinev1.VirtualizedTrustedPlatformModulePolicyEnabled {
```

Reply:
Absolutely, thank you for spotting this :+1:  I have updated this line.